ably/auth.go

```diff
@@ -42,21 +63,87 @@ func (c Capability) String() string {
 }
 
 type Token struct {
-	Token      string     `json:"token"`
-	KeyName    string     `json:"keyName"`
-	Expires    int64      `json:"expires"`
-	Issued     int64      `json:"issued"`
-	Capability Capability `json:"capability"`
+	Token      string     `json:"token" msgpack:"token"`
+	KeyName    string     `json:"keyName" msgpack:"keyName"`
+	Expires    int64      `json:"expires" msgpack:"expires"`
+	Issued     int64      `json:"issued" msgpack:"issued"`
+	Capability Capability `json:"capability" msgpack:"capability"`
+}
+
+var _ msgpack.Unmarshaler = (*Token)(nil)
+
+func (tok *Token) UnmarshalMsgpack(p []byte) error {
+	// This method is workaround for msgpack decoder, which tries to decode fields
+	// using its original type and (finding field byte boundary) and then passes
+	// the bytes to custom unmarshaller.
+	//
+	// Decoding *Token directly fails with:
+	//
+	//   msgpack: invalid code ab decoding map length
+	//
```

Comment:
I suspect the issue is that `Token.Capability` has type `Capability`, but you have defined the custom `UnmarshalMsgpack` on `*Capability`, so msgpack is using the default unmarshaler for a `map[string][]string` but passing it an encoded JSON string (as that is what `Capability.MarshalMsgpack` returns).

I'm pretty sure if you just drop the custom marshaler methods on `Capability`, this will just work.